### PR TITLE
debug: option to omit goroutine stacks by default from debug zip

### LINKED ISSUE #110103: debug zip: make option to "opt out" of stack dumps
**Is your feature request related to a problem? Please describe.**
The `/debug/stacks/<node_id>` endpoint called by debug zip is a heavyweight, "stop the world" operation that uses the `runtime.Stack()` function. Stopping the world on a production cluster can lead to large momentary spikes in SQL Service Latency, which is disruptive especially on larger clusters/workloads.

**Describe the solution you'd like**
Create an option to avoid the pulling of the `/debug/stacks/<node_id>` dumps made by `debug zip`. This will give users the option to take a debug zip with a minimized performance impact.

Jira issue: CRDB-31269

### PR DESCRIPTION
**Pull request by abarganier**
Recently, customers have complained about the performance
impact of debug zip. Specifically, the `/debug/stacks/<node_id>`
endpoint has been found to cause significant momentary spikes
in SQL Service Latency for the node serving the request.

This is because the endpoint uses `runtime.Stack()` from the
go runtime to service the request, which is a "stop-the-world"
operation. Naturally, this degrades performance "while the world
is stopped", which larger customers have found to be unacceptable.

This patch adds the `--include-goroutine-stacks` flag to debug zip,
which defaults to `true`. This gives the option to avoid fetching
goroutine stacks from `/debug/stacks/<node_id>`. This enables
customers to reduce the performance impact of taking a debug zip
bundle.

Release note (ops change): The `debug zip` command now has an
option to omit goroutine stack dumps by. This impacts the creation
of `nodes/*/stacks.txt` and `nodes/*/stacks_with_labels.txt` within
debug zip bundles. Users can opt to exclude these goroutine
stacks by using the `--include-goroutine-stacks=false` flag.

Note that fetching stack traces for all goroutines is a
"stop-the-world" operation, which can momentarily have negative
impacts on SQL service latency. Note that any periodic goroutine
dumps previously taken on the node will still be included in
`nodes/*/goroutines/*.txt.gz`, as these would have already been
generated and don't require any stop-the-world operations.

Fixes: https://github.com/cockroachdb/cockroach/issues/110103

### CODE REVIEW

**Review by dhartunian — COMMENTED** [review]
:lgtm: assuming you change to on by default as discussed. :smile: 

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110177#-:-NdkR-ig51vQ-pxhpO4V:bxow6zg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier and @Santamaura)

___
*[`pkg/cli/context.go` line 384 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110177#-NdkR7Wy-cz7B4Pwr_bl:-NdkR7Wy-cz7B4Pwr_bm:b-3lefdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3066f30152d36d485b45da33d9c8b572d8646750/pkg/cli/context.go#L384)):*
> ```Go
> 	// which impacts performance and SQL service latency. Therefore, we make it opt-in,
> 	// instead of opt-out.
> 	zipCtx.includeStacks = false
> ```

change to `true`.

**Review by abarganier — COMMENTED** [review]
TFTR! Just changed from "opt-in" to "opt-out" as discussed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110177#-:-Ndkhko7BWzN0q4b2B2U:b6jwq3h)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Santamaura)

File: pkg/cli/cliflags/flags.go
Comment (dt) [inline pkg/cli/cliflags/flags.go:1663]: what makes us say that the goroutine dumps previously taken and sitting on disk  `don't require any stop-the-world operations.` ? those call the same `runtime.Stack(..., true)` don't they?

Comment (abarganier) [reply pkg/cli/cliflags/flags.go:1663]: This is meant to say that they don't trigger any "stop the world" operations during the debug zip operation. We just pull the files.

The goroutine dumper uses `pprof`, not `runtime.Stack(..., true)`: https://github.com/cockroachdb/cockroach/blob/4640c585e7098c53ccc69f6096e8087b6d6007fd/pkg/server/goroutinedumper/goroutinedumper.go#L170

Comment (dt) [reply pkg/cli/cliflags/flags.go:1663]: Ah, I see. Perhaps this should say "don't require any _additional_ stop-the-world operations to be collected from disk".

> The goroutine dumper uses pprof, not `runtime.Stack`

https://github.com/golang/go/blob/d80f3966b8b5b2a6e606b416fe80dffeb810ab68/src/runtime/pprof/pprof.go#L703

**Comment by j82w** [comment]
*[`-- commits` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110177#-Ndl7hBzDeXO-5DlskO_):*
Is it possible to do a check at the start of the debug zip and give a warning to disable the flag if the first node has a large number of goroutines? That way all future users will get this warning and won't learn about it until after it causes a big impact to the cluster. It would be nice if we could do this for other flags like nodes too.

File: pkg/cli/cliflags/flags.go
Comment (abarganier) [reply pkg/cli/cliflags/flags.go]: > The goroutine dumper uses pprof, not runtime.Stack

> > https://github.com/golang/go/blob/d80f3966b8b5b2a6e606b416fe80dffeb810ab68/src/runtime/pprof/pprof.go#L703

Lol, well there ya go! 

Pushed a patch to fix the phrasing, as requested.

**Comment by abarganier** [comment]
> Is it possible to do a check at the start of the debug zip and give a warning to disable the flag if the first node has a large number of goroutines? That way all future users will get this warning and won't learn about it until after it causes a big impact to the cluster. It would be nice if we could do this for other flags like nodes too.

@j82w it's possible - we'd have to query TSDB or something like that on each node to make it happen. This would be beyond the effort the team originally intended to put into this, but you're welcome to file an issue and I can tag with the appropriate epic.

**Comment by abarganier** [comment]
TFTRs! 

bors r=dhartunian

**Comment by abarganier** [comment]
Ah, @dhartunian I think we still need a final approval 😅

**Review by dt — APPROVED** [review]

**Review by dhartunian — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110177#-:-NdlS_2ABM4Fp_zJnUgf:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier, @dt, and @Santamaura)

**Comment by abarganier** [comment]
TFTRs! 

bors r=dt,dhartunian

**Pull request merged**
